# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [17]:
import pandas as pd
import numpy as np

In [18]:
url="https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df= pd.read_csv(url)


In [19]:
#Identify the dimensions of the dataset by determining the number of rows and columns it contains.
df.shape
#There are 11 columns and 4008 rows.

(4008, 11)

In [20]:
display(df)

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

In [21]:
#Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. 
#You should also provide suggestions for fixing any incorrect data types.
df.dtypes
#Customer Lifetime Value and Number of Open Complaints should be numerical values.

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

In [22]:
#Want to convert values in this column to floats bbut cannot because the values include non-numerical characters such as %.
#Tested by removing '%' to see if that was the only non-numerical character, it was.
df["Customer Lifetime Value"] = df["Customer Lifetime Value"].str.replace('%', '')
df["Customer Lifetime Value"]=df["Customer Lifetime Value"].astype(float)


In [23]:
df["Customer Lifetime Value"] = df["Customer Lifetime Value"]/100

In [24]:
#Not possible to convert Number of open complaints to a numerical value because the values are formatting as 1/00/00 or 1/2/00 (for example).
#We do not have enought information to understand how to interpret this.
#df["Number of Open Complaints"]=df["Number of Open Complaints"].astype(float)

In [25]:
#Identify the number of unique values for each column and determine which columns appear to be categorical. 
#You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
df.nunique()
#It appears that ST(state), Gender, Education, Policy Type, and Vehicle Class are categorical columns. 
#Given the low number of unique values, it looks at first glance that Number of Open Complaints is also categorical, 
#however it is more likely that the total number of open complaints ranges from zero to five.


Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

In [26]:
#For cleanliness, all column headers should be changed to lowercase and spaces should be changed to '_'.

In [27]:
#Unique values ST column:
unique_st_values = df['ST'].unique()
unique_st_values
#To clean this we need to transform WA to Washington, AZ to Arizona, and Cali to California.

array(['Washington', 'Arizona', 'Nevada', 'California', 'Oregon', 'Cali',
       'AZ', 'WA', nan], dtype=object)

In [28]:
#Unique values Gender column:
unique_gender_values = df['GENDER'].unique()
unique_gender_values
#To clean this, instances of Femal and female should be changed to F and instances fo Male should be chaged to M.

array([nan, 'F', 'M', 'Femal', 'Male', 'female'], dtype=object)

In [29]:
#Unique values Gender column:
unique_education_values = df['Education'].unique()
unique_education_values
#To clean this, instances of Bachelors should be changed to Bachelor, and Master' should be changed to Master.

array(['Master', 'Bachelor', 'High School or Below', 'College',
       'Bachelors', 'Doctor', nan], dtype=object)

In [30]:
#Unique values Gender column:
unique_policy_type_values = df['Policy Type'].unique()
unique_policy_type_values
#Looks clean

array(['Personal Auto', 'Corporate Auto', 'Special Auto', nan],
      dtype=object)

In [31]:
#Unique values Gender column:
unique_vehicle_class_values = df['Vehicle Class'].unique()
unique_vehicle_class_values
#Looks clean

array(['Four-Door Car', 'Two-Door Car', 'SUV', 'Luxury SUV', 'Sports Car',
       'Luxury Car', nan], dtype=object)

In [32]:
#Unique values Number of Open Complaints column:
unique_num_comp_values = df['Number of Open Complaints'].unique()
unique_num_comp_values
#It looks like the second number (after the first "/") is the actual number of complaints. Would need to clarify this assumption.

array(['1/0/00', '1/2/00', '1/1/00', '1/3/00', '1/5/00', '1/4/00', nan],
      dtype=object)

In [33]:
#Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. 
#You should also provide your conclusions based on these summary statistics.
#There is something funky with Customer Lifetime value - the values are percentages, but they are very high. 
#If we assume there was a conversion error in generating the initial data set here, we can divide the values by 100 and the outcome looks more realistic.
#Average CLV is $7,937.
#Average Income is $39,296 - I assume this is annual. 
#Average monthly auto premium is $193.
#Average total claim amount is $405. This is highly skewed by a few large values.

df.describe()

Customer Lifetime Value        Income  Monthly Premium Auto  \
count              1068.000000   1071.000000           1071.000000   
mean               7936.902593  39295.701214            193.234360   
std                6434.784103  30469.427060           1601.190369   
min                2004.350700      0.000000             61.000000   
25%                4034.079950  14072.000000             68.000000   
50%                5881.742350  36234.000000             83.000000   
75%                8962.871975  64631.000000            109.500000   
max               58166.553500  99960.000000          35354.000000   

       Total Claim Amount  
count         1071.000000  
mean           404.986909  
std            293.027260  
min              0.382107  
25%            202.157702  
50%            354.729129  
75%            532.800000  
max           2893.239678

In [34]:
#Compute summary statistics for categorical columns and providing your conclusions based on these statistics.
df.describe(include="object")
#The majority of policy holders are women. The mjority are in Oregon. The majority hold a Bacholor degree. The mjority have personal auto insurance. The majority have four-door cars.
#These values might be different once the data is cleaned as noted above.

Customer      ST GENDER Education Number of Open Complaints  \
count      1071    1071    954      1071                      1071   
unique     1071       8      5         6                         6   
top     RB50392  Oregon      F  Bachelor                    1/0/00   
freq          1     320    457       324                       830   

          Policy Type  Vehicle Class  
count            1071           1071  
unique              3              6  
top     Personal Auto  Four-Door Car  
freq              780            576

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [35]:
counts_per_value = df['ST'].value_counts()


In [36]:
st_series=pd.Series(df['ST'])
st_count_per_value = st_series.value_counts().sort_values(ascending=True)
print(st_count_per_value.head(5))

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64


### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [37]:
policy_series=pd.Series(df['Policy Type'])
policy_count_per_value = policy_series.value_counts().sort_values(ascending=False)
print(policy_count_per_value.head(1))

Policy Type
Personal Auto    780
Name: count, dtype: int64


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [38]:
df_personal= df.loc[df["Policy Type"]=="Personal Auto"]
df_corporate= df.loc[df["Policy Type"]=="Corporate Auto"]

In [39]:
income_perso=round(df_personal["Income"].mean(),0)
income_corporate=round(df_corporate["Income"].mean(),0)
print("Average income for personal policies: " + str(income_perso))
print("Average income for corporate policies: " + str(income_corporate))
#Average income for corporate policies is higher.


Average income for personal policies: 38181.0
Average income for corporate policies: 41390.0


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [40]:
top_quartile_value = df['Total Claim Amount'].quantile(0.75)
total_claim_top_25_df = df[df['Total Claim Amount'] >= top_quartile_value]
display(total_claim_top_25_df)

Customer          ST GENDER Education  Customer Lifetime Value   Income  \
1     QZ44356     Arizona      F  Bachelor                6979.5359      0.0   
2     AI49188      Nevada      F  Bachelor               12887.4317  48767.0   
17    OE15005        Cali    NaN   College                3945.2416  28855.0   
23    TZ98966      Nevada    NaN  Bachelor                2450.1910      0.0   
26    US89481  California    NaN  Bachelor                3946.3721      0.0   
...       ...         ...    ...       ...                      ...      ...   
1059  YG44474      Oregon      M   College               14014.7213  54193.0   
1061  RY92647        Cali      F  Bachelor               10506.7717      0.0   
1068  GS98873     Arizona      F  Bachelor                3239.1247  16061.0   
1069  CW49887  California      F    Master                4626.8011  79487.0   
1070  MY31220  California      F   College                8997.0402  54230.0   

      Monthly Premium Auto Number of Open Complaints     Policy Type  \
1                     94.0                    1/0/00   Personal Auto   
2                    108.0                    1/0/00   Personal Auto   
17                   101.0                    1/0/00   Personal Auto   
23                    73.0                    1/3/00  Corporate Auto   
26                   111.0                    1/0/00   Personal Auto   
...                    ...                       ...             ...   
1059                 117.0                    1/0/00  Corporate Auto   
1061                  92.0                    1/0/00   Personal Auto   
1068                  88.0                    1/0/00   Personal Auto   
1069                 114.0                    1/0/00    Special Auto   
1070                 112.0                    1/0/00   Personal Auto   

      Vehicle Class  Total Claim Amount  
1     Four-Door Car         1131.464935  
2      Two-Door Car          566.472247  
17              SUV          647.442031  
23    Four-Door Car          554.376763  
26    Four-Door Car          799.200000  
...             ...                 ...  
1059            SUV          720.752945  
1061  Four-Door Car          546.524896  
1068  Four-Door Car          633.600000  
1069            SUV          547.200000  
1070   Two-Door Car          537.600000  

[273 rows x 11 columns]

In [41]:
total_claim_top_25_df.describe()

Customer Lifetime Value        Income  Monthly Premium Auto  \
count               273.000000    273.000000            273.000000   
mean               9183.794161  24106.545788            163.000000   
std                6582.581975  27542.717560            613.645376   
min                2287.596900      0.000000             63.000000   
25%                4747.734600      0.000000             99.000000   
50%                7859.414600  18846.000000            113.000000   
75%               10656.882000  44897.000000            133.000000   
max               40224.013600  99316.000000          10202.000000   

       Total Claim Amount  
count          273.000000  
mean           774.005353  
std            291.304731  
min            532.800000  
25%            595.200000  
50%            673.342650  
75%            844.481918  
max           2893.239678

In [42]:
total_claim_top_25_df.describe(include="object")

Customer      ST GENDER             Education  \
count       273     273    245                   273   
unique      273       7      5                     5   
top     QZ44356  Oregon      F  High School or Below   
freq          1      93    118                    90   

       Number of Open Complaints    Policy Type Vehicle Class  
count                        273            273           273  
unique                         6              3             6  
top                       1/0/00  Personal Auto           SUV  
freq                         213            195           105

There are 273 policies with a claim amount greater than the 75th percentile. On average these customers pay $163/month, earn an income of $24,107/year and have a CLV of $9,184.
The majority of policies in this group are in Oregon, are women, have a high school education (or lower), are for an SUV, and are for personal insurance.